typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs

```diff
@@ -197,6 +207,45 @@ channel_experiment clientChannel serverChannel (Positive x) probe = do
   res <- atomically $ takeTMVar serverVar
   probeOutput probe (res === c)
 
+-- | An existential type which contains @PeerHasAgency pk st@ singleton and
+-- a message from the @st@ state.
+--
+data AnyPingPongMessage
+  = forall (pk :: PeerKind) st st'. AnyPingPongMessage (PeerHasAgency pk st) (Message PingPong st st')
+
+instance Arbitrary AnyPingPongMessage where
+  arbitrary = oneof
+    [ return (AnyPingPongMessage (ClientAgency TokIdle) MsgPing)
+    , return (AnyPingPongMessage (ServerAgency TokBusy) MsgPong)
+    , return (AnyPingPongMessage (ClientAgency TokIdle) MsgDone)
+    ]
+
+instance Show AnyPingPongMessage where
+  show (AnyPingPongMessage _ msg) = show $ "PingPongClientMessage " ++ show msg
+
+prop_codec
+  :: AnyPingPongMessage
```

Comment:
Is it feasible to make a generic codec consistency property? Give a generator for some transition type, a codec for it, and get a property?

Reply:
I thought about it too, but I didn't try. It would need to get three things:
* codec
* `otherPeer :: PeerHasAgency (pk :: PeerKind) (st :: PingPong) -> Proxy (FlipAgency pk)`
* and `eqMsg :: Message ps st st' -> Message ps st st' -> Bool`

Reply:
Right you'd in fact need to hide the transition endpoints:

```Haskell
eqMsg :: SomeMessage ps -> SomeMessage ps -> Maybe NotEqualExplanation
```

which is fine. As for `otherPeer`, why should that be needed? Seems this test will run the encoder and decoder at different `tok` values, which strikes me as odd. I would have thought that the property we want is: the decoder inverts the encoder _at a particular state_ (the same state).

Reply:
Indeed, it's done. I went with an `Eq (AnyMessage ps)` instance (here `AnyMessage` is isomorphic to your `SomeMessage`; `SomeMessage` was taken).